src/kzg10/subproductdomain.rs

```diff
@@ -0,0 +1,424 @@
+use crate::Error;
+use ark_ec::PairingEngine;
+use ark_ff::{FftField, Field, Zero};
+use ark_poly::{Polynomial, UVPolynomial};
+
+use ark_poly::polynomial::univariate::DensePolynomial as Poly;
+
+/// Where indicated, algorithms are from Modern Computer Algebra, 3rd edition, by Gathen and Gerhard
+/// Abbreviated as GG
+/// Let M(n) denote the time to multiply.
+
+/// GG Algorithm 9.3
+/// Computes the inverse of f mod x^l
+/// Takes O(M(l)) field arithmetic operations
+pub fn inverse_mod_xl<F: FftField>(f: &Poly<F>, l: usize) -> Option<Poly<F>> {
+    let r = ark_std::log2(l); // Compute ceil(log_2(l))
+    if let Some(f_0_inv) = f.coeffs[0].inverse() {
+        // Invert f(0)^-1 if possible
+        let mut g = Poly::<F> {
+            coeffs: vec![f_0_inv], // Constant polynomial f(0)^-1
+        };
+
+        let mut i = 2usize;
+        // Use Newton iteration which converges to the inverse mod x^l
+        for _ in 0..r {
+            g = &(&g + &g) - &(f * &(&g * &g)); //TODO: is g*2 better than g+g?
```

Comment:
You can do something like `g.double()` and `g.square()`. These should be slightly faster than computing the addition and multiplication directly.

Reply:
Oh, that makes sense, I hadn't realized there were API calls for this